designs/simplified-scaffolding.md

```diff
@@ -0,0 +1,512 @@
+to pass in dependencies from `main`.
+
+### Why don't we...
+
+#### Put the controller setup code in main.go
+
+While this is an attractive pattern from a prototyping perspective, it
+makes it harder to write integration tests, since you can't easily say
+"run this controller with all its setup in processes".  With a separate
+`SetupWithManager` method associated with reconcile, it becomes fairly
+easy to setup with a manager.
+
+#### Put the types directly under api/, or not have groupversion_info.go
+
+These suggestions make it much harder to scaffold out additional versions
+and kinds.  You need to have each version in a separate package, so that
+type names don't confict.  While we could put scheme registration in with
+`kind_types.go`, if a project has multiple "significant" Kinds in an API
+group, it's not immediately clear which file has the scheme registration.
+
+#### Use a single types.go file
+
+This works fine when you have a single "major" Kind, but quickly grows
+unweildy when you have multiple major kinds and end up with
+a hundreds-of-lines-long `types.go` file (e.g. the `appsv1` API group in
+core Kubernetes).  Splitting out by "major" Kind (`Deployment`,
+`ReplicaSet`, etc) makes the code organization clearer.
+
+#### Change the current scaffold to just make Add a method on the reconciler
+
+While this solves the dependency issues (mostly, since you might want to
+further pass configuration to the setup logic and not just the runtime
+logic), it does not solve the underlying pedagogical issues around the
+initial structure burying key logic amidst a sprawl of generated files and
+directories.
+
+### Making this work with multiple controllers, API versions, API groups, etc
+
+#### Versions
+
+Most projects will eventually grow multiple API versions.  The only
+wrinkle here is making sure API versions get added to a scheme.  This can
+be solved by adding a specially-marked init function that registration
+functions get added to (see the example).
+
+#### Groups
+
+Some projects eventually grow multiple API groups.  Presumably, in the
+case of multiple API groups, the desired hierarchy is:
+
+```shell
+$ tree ./test/project/api
+./test/project/api
+├── groupa
+│   └── v1
+│       └── types.go
+└── groupb
+    └── v1
+        └── types.go
+```
+
+There are three options here:
+
+1. Scaffold with the more complex API structure (this looks pretty close
+   to what we do today).  It doesn't add a ton of complexity, but does
+   bury types deeper in a directory structure.
+
+2. Try to move things and rename references.  This takes a lot more effort
+   on the KubeBuilder maintainers' part if we try to rename references
+   across the codebase.  Not so much if we force the user to, but that's
+   a poorer experience.
+
+3. Tell users to move things, and scaffold out with the new structure.
+   This is fairly messy for the user.
+
+Since growing to multiple API groups seems to be fairly uncommon, it's
```

Comment:
> fairly uncommon

IIUC you mentioned that you have some data to back this when we chat offline. If you put the number here it's more convincing to make the decision :)

Reply:
swear I had this somewhere, but can't seem to find the data now.  I'll see what I can put together

Reply:
hmm... reproducing what I had before is proving difficult (the query I was using before is now returning wildly truncated results), but with a random sampling of smaller projects on github using kubebuilder (presence of manager.yaml) and larger projects that I know of that make use of controller runtime or just CRDs in general (e.g. kubeflow, knative, cluster-api, cert-manager, app-crd, prometheus operator, etcd operator) most seem to follow the pattern of 1 API group per repo.  The main exception was a couple of the knative repositories (but not all of them) and openshift, which puts all of its types in a single repo anyway, so isn't relevant to us.

This makes some amount of sense -- most operators or controllers logically operate on a single set of resources (the CRD defining the operator, a collection of related resources that work together) which would generally be grouped into a single API group.

Reply:
hmm...another crazy idea :)

What if we removed the `api` and introduced the *group* hierarchy.

```shell
./test/project
  - groupa/v1/....
  - groupb/v2/...
```